core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java

```diff
@@ -282,7 +279,7 @@ protected OperationBuffer() {
      * @param input the input to operate on
      * @throws IOException if a non-Alluxio error occurs
      */
-    public void add(T input) throws IOException {
+    public synchronized void add(T input) throws IOException {
       if (mCurrentBatchBuffer.size() == getBatchSize()) {
```

Comment:
i feel `==` is vulnerable to bugs, like if for some unexpected reasons the size is already larger, or the submit at `==` did not succeed, this submission might not be triggered again. I'm not saying the current logic doesn't work, I'm just saying any code piece can break in ways we don't anticipate, at any time.

@lucyge2022 do you think it's more ideal if we compare `>=` here? When we `submitBatch()`, the UFS may have a request size limit (so you cannot submit more than maybe 1000 in the batch). In that case, we should submit the 1st 1000 items and leave the rest in the buffer. Essentially in the `submitBatch()` it won't be a copy using `new ArrayList<>(buffer)`, but something like `for (i < 1000) { list.add(buffer.poll() }`

Reply:
we could utilize arrayblockingq's capacity for mCurrentBatchBuffer, if we can't add more which means its time for submit batch, and when submiting batch we could use drainto to drain the added elements and submit to threadpool for batch processing.